beacon-chain/core/state/validity_conditions.go

```diff
@@ -0,0 +1,181 @@
+	"github.com/prysmaticlabs/prysm/shared/bitutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// IsValidBlock verifies a block is valid according to the ETH 2.0 specification for
+// validity conditions taking into consideration attestation processing and more.
+func IsValidBlock(
+	block *types.Block,
+	beaconState *types.BeaconState,
+	parentSlot uint64,
+	genesisTime time.Time,
+	isInChain func(blockHash [32]byte) bool,
+) error {
+	_, err := block.Hash()
+	if err != nil {
+		return fmt.Errorf("could not hash incoming block: %v", err)
+	}
+
+	if block.SlotNumber() == 0 {
+		return errors.New("cannot process a genesis block: received block with slot 0")
+	}
+
+	if !block.IsSlotValid(genesisTime) {
+		return fmt.Errorf("slot of block is too high: %d", block.SlotNumber())
+	}
+
+	if err := doesParentProposerExist(block, beaconState, parentSlot); err != nil {
+		return fmt.Errorf("could not get proposer index: %v", err)
+	}
+
+	for _, attestation := range block.Attestations() {
+		if err := isBlockAttestationValid(block, attestation, beaconState, parentSlot, isInChain); err != nil {
+			return fmt.Errorf("invalid block attestation: %v", err)
+		}
+	}
+
+	_, proposerIndex, err := v.ProposerShardAndIndex(
+		beaconState.ShardAndCommitteesForSlots(),
+		beaconState.LastStateRecalculationSlot(),
+		block.SlotNumber(),
+	)
+	if err != nil {
+		return fmt.Errorf("could not get proposer index: %v", err)
+	}
+
+	stateProposerRandaoSeed := beaconState.Validators()[proposerIndex].RandaoCommitment
+	blockRandaoReveal := block.RandaoReveal()
+
+	// If this is a block created by the simulator service (while in development
+	// mode), we skip the RANDAO validation condition.
+	isSimulatedBlock := bytes.Equal(blockRandaoReveal[:], params.BeaconConfig().SimulatedBlockRandao[:])
+	if !isSimulatedBlock && !block.IsRandaoValid(stateProposerRandaoSeed) {
+		return fmt.Errorf(
+			"pre-image of %#x is %#x, Got: %#x",
+			blockRandaoReveal[:],
+			hashutil.Hash(blockRandaoReveal[:]),
+			stateProposerRandaoSeed,
+		)
+	}
+	return nil
+}
+
+// doesParentProposerExist checks that the proposer from the parent slot is included in the first
+// aggregated attestation object
+func doesParentProposerExist(block *types.Block, beaconState *types.BeaconState, parentSlot uint64) error {
+	_, parentProposerIndex, err := v.ProposerShardAndIndex(
+		beaconState.ShardAndCommitteesForSlots(),
+		beaconState.LastStateRecalculationSlot(),
+		parentSlot,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Verifies the attester bitfield to check if the proposer index is in the first included one.
+	if isBitSet, err := bitutil.CheckBit(block.Attestations()[0].AttesterBitfield, int(parentProposerIndex)); !isBitSet {
+		return fmt.Errorf("could not locate proposer in the first attestation of AttestionRecord: %v", err)
+	}
+	return nil
+}
+
+// isBlockAttestationValid verifies a block's attestations pass validity conditions.
```

Comment:
Likewise here, it would be nice to see a summary of the validation conditions in this comment.